File: app/pages/repos/_repo_id/models/_model_id/training_importances/props.ts
import { ModelLayoutInfo } from "layouts/model_layout"

export type Props = {
	inner: Inner
	modelLayoutInfo: ModelLayoutInfo
}

export type Inner =
	| {
			type: Type.LinearRegressor
			value: LinearRegressorProps
	  }
	| {
			type: Type.LinearBinaryClassifier
			value: LinearBinaryClassifierProps
	  }
	| {
			type: Type.LinearMulticlassClassifier
			value: LinearMulticlassClassifierProps
	  }
	| {
			type: Type.TreeRegressor
			value: TreeRegressorProps
	  }
	| {
			type: Type.TreeBinaryClassifier
			value: TreeBinaryClassifierProps
	  }
	| {
			type: Type.TreeMulticlassClassifier
			value: TreeMulticlassClassifierProps
	  }

export enum Type {
	LinearRegressor = "linear_regressor",
	TreeRegressor = "tree_regressor",
	LinearBinaryClassifier = "linear_binary_classifier",
	TreeBinaryClassifier = "tree_binary_classifier",
	LinearMulticlassClassifier = "linear_multiclass_classifier",
	TreeMulticlassClassifier = "tree_multiclass_classifier",
}

export type FeatureImportance = {
	featureImportanceValue: number
	featureName: string
}

export type LinearRegressorProps = {
	featureImportances: FeatureImportance[]
	nFeatures: number
}

export type LinearBinaryClassifierProps = {
	featureImportances: FeatureImportance[]
	nFeatures: number
}

export type LinearMulticlassClassifierProps = {
	featureImportances: FeatureImportance[]
	nFeatures: number
}

export type TreeRegressorProps = {
	featureImportances: FeatureImportance[]
	nFeatures: number
}

export type TreeBinaryClassifierProps = {
	featureImportances: FeatureImportance[]
	nFeatures: number
}

export type TreeMulticlassClassifierProps = {
	featureImportances: FeatureImportance[]
	nFeatures: number
}

File: app/pages/repos/_repo_id/models/_model_id/training_importances/props.rs
use tangram_app_layouts::model_layout::ModelLayoutInfo;

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Props {
	pub id: String,
	pub inner: Inner,
	pub model_layout_info: ModelLayoutInfo,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "type", content = "value")]
pub enum Inner {
	LinearRegressor(LinearRegressorProps),
	TreeRegressor(TreeRegressorProps),
	LinearBinaryClassifier(LinearBinaryClassifierProps),
	TreeBinaryClassifier(TreeBinaryClassifierProps),
	LinearMulticlassClassifier(LinearMulticlassClassifierProps),
	TreeMulticlassClassifier(TreeMulticlassClassifierProps),
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FeatureImportance {
	pub feature_importance_value: f32,
	pub feature_name: String,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LinearRegressorProps {
	pub feature_importances: Vec<FeatureImportance>,
	pub n_features: usize,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TreeRegressorProps {
	pub feature_importances: Vec<FeatureImportance>,
	pub n_features: usize,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LinearBinaryClassifierProps {
	pub feature_importances: Vec<FeatureImportance>,
	pub n_features: usize,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TreeBinaryClassifierProps {
	pub feature_importances: Vec<FeatureImportance>,
	pub n_features: usize,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LinearMulticlassClassifierProps {
	pub feature_importances: Vec<FeatureImportance>,
	pub n_features: usize,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TreeMulticlassClassifierProps {
	pub feature_importances: Vec<FeatureImportance>,
	pub n_features: usize,
}
